cmd/audius-ctl/status.go

```diff
@@ -2,45 +2,308 @@ package main
 
 import (
 	"fmt"
+	"os"
+	"sort"
+	"sync"
+	"time"
 
 	"github.com/AudiusProject/audius-d/pkg/conf"
+	"github.com/AudiusProject/audius-d/pkg/health"
 	"github.com/AudiusProject/audius-d/pkg/logger"
-	"github.com/AudiusProject/audius-d/pkg/test"
+	"github.com/jedib0t/go-pretty/v6/table"
+	"github.com/jedib0t/go-pretty/v6/text"
 	"github.com/spf13/cobra"
 )
 
+type hcResult struct {
+	Host          string
+	HealthSummary health.NodeHealthSummary
+	Error         error
+}
+
+type diskUsage struct {
+	sizeBytes uint64
+	usedBytes uint64
+}
+
+const (
+	nodeCol int = iota
+	typeCol
+	upCol
+	healthyCol
+	chainCol
+	websocketCol
+	ipCol
+	dbCol
+	diskCol
+	uptimeCol
+	commentCol
+)
+
+const (
+	diskUsageWarningThreshold float64 = 0.80
+	diskUsageErrorThreshold   float64 = 0.90
+	dbSizeWarningThreshold    uint64  = 2
+)
+
 var (
-	testCmd = &cobra.Command{
-		Use:          "status",
-		Short:        "test audius-d connectivity",
+	noStatus     = "n/a"
+	ignoreHealth bool
+	statusCmd    = &cobra.Command{
+		Use:          "status [host ...]",
+		Short:        "Check health of configured nodes",
 		SilenceUsage: true, // do not print --help text on failed node health
-		Args:         cobra.ExactArgs(0),
 		RunE: func(cmd *cobra.Command, args []string) error {
 			ctxConfig, err := conf.ReadOrCreateContextConfig()
-			if err != nil {
-				return logger.Error("Failed to retrieve context. ", err)
-			}
-
-			responses, err := test.CheckNodeHealth(ctxConfig)
 			if err != nil {
 				return err
 			}
 
-			var encounteredError bool
-			for _, response := range responses {
-				if response.Error != nil {
-					fmt.Printf("%-50s Error: %v\n", response.Host, response.Error)
-					encounteredError = true
-				} else {
-					fmt.Printf("%-50s %t\n", response.Host, response.Result)
+			var nodesToCheck map[string]conf.NodeConfig
+			if len(args) == 0 {
+				nodesToCheck = ctxConfig.Nodes
+			} else {
+				nodesToCheck, err = filterNodesFromContext(args, ctxConfig)
+				if err != nil {
+					return err
 				}
 			}
 
-			if encounteredError {
-				return fmt.Errorf("\none or more health checks failed")
+			var wg sync.WaitGroup
+			resultsChan := make(chan hcResult, len(nodesToCheck))
+			for host, config := range nodesToCheck {
+				wg.Add(1)
+				go func(h string, c conf.NodeConfig) {
+					defer wg.Done()
+					result, err := health.CheckNodeHealth(h, c)
+					resultsChan <- hcResult{
+						Host:          h,
+						HealthSummary: result,
+						Error:         err,
+					}
+				}(host, config)
 			}
 
-			return nil
+			go func() {
+				wg.Wait()
+				close(resultsChan)
+			}()
+
+			var results []hcResult
+			for r := range resultsChan {
+				results = append(results, r)
+			}
```

Comment:
interesting waitgroup usage

Reply:
i don't think you _need_ the results channel and you can just wait the wg in the main thread, add to the results straight from the goroutine. not necessary too

Reply:
I think I'll leave it for now, it's not thread safe to append to a slice, and while we could technically add to the results array by index, I don't think it's a necessary optimization. (https://stackoverflow.com/questions/44152988/append-not-thread-safe)